### PR TITLE
Setup wizard: Done screen

### LINKED ISSUE #4707: Wizard step: "Done" screen
## Feature description

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Display headline about branding your theme
* Let users know they're about to be sent to the Customizer to see how their site looks with AMP enabled (Customizer AMP toggle turned on)
* Next button and back button
* Next button takes you to the Customizer

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by johnwatkins0**
## Summary

Fixes #4707

This adds the final screen for the setup wizard. Uses are presented with an iframe showing an AMP page on their site in a mobile viewport, along with links to view the site or customize (in Reader mode). The "Next" button becomes "Finish" and links back to AMP settings.

Screen Shot 2020-06-29 at 11 04 03 PM

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [inline assets/src/setup/pages/save/index.js]:
```suggestion
						sandbox="allow-scripts allow-forms allow-popups allow-presentation"
```

**Comment by johnwatkins0** [comment]
@westonruter This PR is still WIP, but I wanted to flag for you that I made an attempt to hide the admin bar in the iframe using the `sandbox` attribute. I expected it to work, but for some reason it didn't. I could sense that figuring this out was potentially going to take a signicant amount of time, so I went to the fallback of setting a `amp-no-admin-bar` URL var and turning off the admin bar when it is set. 

This PR still has the iframe with the `sandbox` parameter. To replicate, remove the `show_admin_bar` filter I added in `includes/amp-helper-functions.php`. Then, in the wizard, select standard mode and navigate to the last screen, which has the iframe. Where we would expect the iframe to be hidden, it's still showing.

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: You got this in just ahead of my comment below. I'll try it in a little while.

**Comment by westonruter** [comment]
I may have been mistaken on the behavior of `sandbox` and whether it blocks cookies from being used then. Maybe it's just a matter of whether the iframed window can read cookies from the parent window or not.

**Comment by westonruter** [comment]
I have another idea: provide a `name` to the `iframe`, like `amp-wizard-completion-preview`. Then, you could do something like this:

```php
add_action( 'wp_print_footer_scripts', function() {
	if ( ! amp_is_dev_mode() || ! is_admin_bar_showing() ) {
		return;
	}
	?>
	<script data-ampdevmode>
        document.addEventListener( 'DOMContentLoaded', () => {
            if ( document.body.classList.contains( 'admin-bar' ) ) {
                document.body.classList.remove( 'admin-bar' );
                document.documentElement.style.cssText += '; margin-top: 0 !important';
                document.getElementById( 'wpadminbar' ).remove();
            }
        } );
	</script>
	<?php
} );
```

But that would still result in a flash of the admin bar appearing momentarily. 

In any case, this is probably overkill. If the admin bar shows up upon navigation, that's not the end of the world 😄

**Comment by johnwatkins0** [comment]
This is very similar:

```PHP
/**
 * Hide admin bar if the window is inside the setup wizard iframe.
 */
add_action(
	'wp_print_footer_scripts',
	function() {
		if ( ! amp_is_dev_mode() || ! is_admin_bar_showing() ) {
			return;
		}
		?>
		<script data-ampdevmode>
			document.addEventListener( 'DOMContentLoaded', function() {
				if (!('name' in window) || 'amp-wizard-completion-preview' !== window.name) { // Detects whether the current window is in an iframe.
					return;
				}

				if ( document.body.classList.contains( 'admin-bar' ) ) {
					document.body.classList.remove( 'admin-bar' );
					document.documentElement.style.cssText += '; margin-top: 0 !important';
					document.getElementById( 'wpadminbar' ).remove();
				}
			});
		</script>
		<?php
	}
);
```

Apparently if a window is in an iframe, the iframe's `name` attribute is available as `window.name`. I've never come across this piece of information before (that I can recall). 

I don't think this is overkill because it's no more complicated and is maybe more reliable than the query var, and works across navigations, so I'm going to add it.

**Comment by johnwatkins0** [comment]
Opening this for review after making a round of CSS updates not necessarily directly related to this task. I also **removed the feature flag** around the setup wizard, so anyone checking out this branch or `develop` after this branch is merged will have access to the setup wizard. No need for the `AMP_NEW_ONBOARDING` constant. This can easily be reversed if I've done it too soon.

**Comment by johnwatkins0** [comment]
@jwold The production build in the above comment should work in your local environment out of the box without having to mess with PHP constants.

**Comment by jwold** [comment]
1. Can we make all the steps blue checkmarks as they get completed? https://d.pr/i/rDAmSj
2. I'm now wondering if we should have the gray outline: https://d.pr/i/ZMfD7t. It is feeling like a button. Curious for your thoughts! cc @amedina and @westonruter

**Comment by johnwatkins0** [comment]
> * Can we make all the steps blue checkmarks as they get completed? https://d.pr/i/rDAmSj
> * I'm now wondering if we should have the gray outline: https://d.pr/i/ZMfD7t. It is feeling like a button. Curious for your thoughts! cc @amedina and @westonruter

Now being handled in #4962

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [inline assets/src/setup/pages/save/index.js]: See comment above about returning the permalink via the REST API.

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [inline assets/src/setup/pages/save/index.js]: The `amp` query var needs to be added when going to the non-Legacy Reader mode Customizer:

```suggestion
											: { url: PREVIEW_PERMALINK, amp: 1 },
```

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [inline assets/src/setup/pages/save/index.js]: I suppose manipulating the query vars is needed because the output of `amp_admin_get_preview_permalink` may now be stale.

Here's an idea: can the REST API endpoint include the preview link URL? Perhaps a new endpoint? Or part of the options endpoint as a read-only value? Then when saving we will get the most up-to-date preview URL based on the settings. This will be very important when we allow sites to customize how the URLs are constructed for AMP.  See #2204.

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [inline assets/src/setup/pages/save/index.js]: See comment below about getting `PREVIEW_PERMALINK` from the REST API.

File: includes/options/views/class-amp-setup-wizard-submenu-page.php
Comment (westonruter) [inline includes/options/views/class-amp-setup-wizard-submenu-page.php]: See comment above about how this should ideally be returned by the REST API after saving the mode.

**Review by westonruter — CHANGES_REQUESTED** [review]

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: I actually considered (1) including the preview link in the options response after saving, and (2) creating a new REST endpoint just for the preview link. 1 feels wrong because it's not an option, and 2 felt like overkill for a single field, but with your suggestion I'll go ahead with 1. It is sort of an option, just not one that persists in the DB.

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: I suppose this won't be needed with the preview permalink returned via the Options endpoint.

Comment (westonruter) [reply assets/src/setup/pages/save/index.js]: It will because this is an `amp` query var for `customize.php` not the `PREVIEW_PERMALINK`.

Comment (westonruter) [reply assets/src/setup/pages/save/index.js]: Granted, at the moment this does nothing. But it _will_.

File: includes/options/class-amp-options-rest-controller.php
Comment (westonruter) [inline includes/options/class-amp-options-rest-controller.php:107]: This is correct anyway, as the option is computed rather than something stored.

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: OK, I hadn't picked up on the customizer link also needing the `amp` query var. Updated.

File: includes/options/views/class-amp-setup-wizard-submenu-page.php
Comment (westonruter) [reply includes/options/views/class-amp-setup-wizard-submenu-page.php]:
```suggestion
```

File: assets/src/setup/pages/save/style.css
Comment (westonruter) [inline assets/src/setup/pages/save/style.css]: This is turning out to be extremely narrow, like 240px:

image

Therefore, I think the iframe needs to _not_ be floated to the right and instead appear _below_ the copy. The iframe should then have a `max-width:640px`.

Comment (westonruter) [reply assets/src/setup/pages/save/style.css]: Or perhaps a `max-width:480px`. Whatever the most common screen resolution is for mobile phones.

Example:

image

cc @jwold

**Comment by jwold** [comment]
Let’s do it. Below is not as nice, but better. Can we just make it drop for smaller screens?

File: assets/src/setup/pages/save/style.css
Comment (johnwatkins0) [reply assets/src/setup/pages/save/style.css]: I just set a fixed width of 400px on that column (approximately what it is on the design) and have it wrapping below a certain width) in a4003579f59bdb0cf23c4b8d3c9bf7f2e2834b29

Screen Shot 2020-06-30 at 4 29 49 PM

**Comment by jwold** [comment]
Looks good to me 
GitHub <notifications@github.com> wrote:
“@johnwatkins0 commented on this pull request.”


“In assets/src/setup/pages/save/style.css:”

> +	text-align: center;
+}
+
+.done__preview-container .phone {
+	max-height: 610px;
+	padding-bottom: 2rem;
+}
+
+.done__preview-container .phone::before {
+	height: 9px;
+	width: 80px;
+}
+
+.done__preview-iframe {
+	height: 610px;
+	width: 100%;
“I just set a fixed width of 400px on that column (approximately what it is on the design) and have it wrapping below a certain width).”

[Screen Shot 2020-06-30 at 4 29 49 PM]
“ —

**Review by westonruter — APPROVED** [review]

**Comment by schlessera** [comment]
How about scaling the iframe content - http://jsfiddle.net/6th5107u/

**Comment by westonruter** [comment]
> How about scaling the iframe content - http://jsfiddle.net/6th5107u/

That's a possibility, but the scaling causes visual regressions (e.g. text legibility):

image

**Comment by schlessera** [comment]
I'm getting an error when trying to complete the wizard:
Image 2020-07-01 at 9 16 05 AM

I made the following choices:
- Developer or technically savvy
- Transitional (Previously selected)
- Astra theme
- Redirect mobile visitors to AMP

**Comment by schlessera** [comment]
I think the bottom bar transparency looks off and is visually inconsistent.
Image 2020-07-01 at 9 22 53 AM
Image 2020-07-01 at 9 23 36 AM

I would personally prefer for this to stick more to the Material principle of everything having a consistent depth and shadows that represent that.
Here's the same bar without translucency and with a subtle shadow applied:
Image 2020-07-01 at 9 22 02 AM
Image 2020-07-01 at 9 22 08 AM

File: assets/src/setup/components/nav/index.js
Comment (pierlon) [inline assets/src/setup/components/nav/index.js:76]: I think we can remove the "next" arrow icon if it's the last page, otherwise it could imply there's an another page to naviagte to.

image

File: assets/src/setup/components/nav/style.css
Comment (pierlon) [inline assets/src/setup/components/nav/style.css:22]: This makes the nav bar look misaligned from the setup wizard:

image

**Comment by johnwatkins0** [comment]
> I'm getting an error when trying to complete the wizard:
> Image 2020-07-01 at 9 16 05 AM

@schlessera This may be because you still have "classic" set as the reader theme, which won't be accepted by the `validate_callback` now that we've switched that slug to `legacy`. I think if you were to select reader mode and choose a reader theme from the list (you would even be able to go back and re-select transitional after choosing a theme) this issue would go away.

**Comment by johnwatkins0** [comment]
@jwold What do you think about @schlessera's suggestion on the bottom nav bar two comments up?

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: WordPress dependency listed under internal.

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: "Finish" button should be disabled when saving. Currently, it's enabled:

image

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: Should we explicitly state that the preview is an AMP page? Otherwise, it may confuse users as to which version they are looking at (especially when in reader/transitional mode):

image


```suggestion
					{ __( 'Live view of your AMP site', 'amp' ) }
```

**Comment by jwold** [comment]
> @jwold What do you think about @schlessera's suggestion on the bottom nav bar two comments up?

Let's go with @schlessera's change. I think it will be an improvement. cc @westonruter and @amedina

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: Should the admin bar be shown when previewing the page with the reader mode template?

image

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: Can the buttons be placed to the left under the description? I don't think most users would be able to see them unless they were to scroll down first:

**Review by pierlon — CHANGES_REQUESTED** [review]

File: assets/src/setup/pages/save/index.js
Comment (amedina) [reply assets/src/setup/pages/save/index.js]: That may be good. Could we try it and see the difference?

Comment (amedina) [reply assets/src/setup/pages/save/index.js]: Also, why do we need a difference between the button for "`Visit in Customizer`" and "`Browse AMP Version`"?

Making them equal will provide symmetry.

**Comment by pierlon** [comment]
Following up from review with 2 questions:

- Should we persist the local "developer tools" and "template mode" options? Otherwise the user has to re-select their chosen options each time if they were to navigate back to the page:

    

- Should the template notice body be clickable?

    image

**Comment by johnwatkins0** [comment]
> > @jwold What do you think about @schlessera's suggestion on the bottom nav bar two comments up?
> 
> Let's go with @schlessera's change. I think it will be an improvement. cc @westonruter and @amedina

Done in edd43cb63

File: assets/src/setup/components/nav/index.js
Comment (johnwatkins0) [reply assets/src/setup/components/nav/index.js:76]: Good idea. Updated in 11adf2dcf

File: assets/src/setup/components/nav/style.css
Comment (johnwatkins0) [reply assets/src/setup/components/nav/style.css:22]: Fixed in d570470a3

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: Done in 05ae9e993f8a3551eb55cf696d839022754ed5d9

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: Here's how that looks. Shall I push it @jwold @amedina ?

Screen Shot 2020-07-01 at 10 47 16 AM

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: I should defer to @westonruter on this one. We have the admin bar removed in the other modes/themes and I'm not immediately sure what differences in the legacy theme might account for this.

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: Updated in 48653bbd1

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: Done (in a less hacky way than my previous attempt) in 48653bbd1

**Comment by johnwatkins0** [comment]
> Following up from review with 2 questions:
> 
> * Should we persist the local "developer tools" and "template mode" options? Otherwise the user has to re-select their chosen options each time if they were to navigate back to the page:

Yes, this is a little annoying when doing QA, but it was agreed in one of our recent calls that we want to have user consciously make these choices when re-running the wizard. 

> * Should the template notice body be clickable?


Updated in 4139e4acd . In 1.7 these buttons will open panels; for now I just added `pointer-events: none` to make them nonclickable.

**Comment by westonruter** [comment]
> Yes, this is a little annoying when doing QA, but it was agreed in one of our recent calls that we want to have user consciously make these choices when re-running the wizard.

The technical user question should persist once the wizard is completed, no?

@johnwatkins0 Also, in regards to the template mode selection, this brings up something I wanted to raise. The template mode should only be unselected (with the "Previously selected" label) if the user _changed_ the value for the technical user question. If I go through the wizard the first time, and I say I'm technical and set the mode to Transitional. The next time I go through the wizard, then the technical option should be pre-selected and if I leave that alone then the Transitional mode will be pre-selected on the next screen as well. _But_ if I change the response to the technical question then at that point no mode should be pre-selected.

**Comment by johnwatkins0** [comment]
@westonruter @pierlon This is ready for review again. This PR has become a catch-all for everything we've discussed for the last couple days, and this afternoon while waiting for builds to run I refactored the e2e tests (to allow us to be able to sequence the tests -- which helps with testing persistent options). Sorry about the extra diff. Let me know if you see any issues.

**Comment by westonruter** [comment]
@johnwatkins0 I went through the wizard after having saved previously with a Reader theme.

1.  The technical question is pre-selected to be non-technical. ✅
2. The Reader mode is selected by default. ✅ 
3. Clicking next (without changing the mode) then takes me _immediately_ to the Summary screen. I should have been able to go to the theme selection. ❌ 
4. The mobile theme screenshot is not populated. ❌ 
5. Clicking Previous takes me to Theme selection but the screen is empty. ❌ 

image

image

To work around this issue I have to make a change to the mode and then switch back to Reader, and then I am taken to the theme selection and see the screenshot as expected on the summary screen.

**Comment by johnwatkins0** [comment]
Thanks, @westonruter . 7ab0495e8cc11798adeabb73644f669c3fac13a5 should fix that. I had to make some changes to the way options are held in the application state, so that now we have `originalOptions` (what's received from the endpoint), `updates` (changed values) and `editedOptions` ( `{ ...originalOptions, ...updates } )`, and it looks I missed switching `updates` to `editedOptions` in a couple places where the latter should be used.

**Comment by westonruter** [comment]
@johnwatkins0 When clicking the Finish button I'm getting a JS error:

image

**Comment by westonruter** [comment]
`activePageIndex` is `5` so and there is no `activePageIndex + 1` page:

image

**Comment by johnwatkins0** [comment]
> @johnwatkins0 When clicking the Finish button I'm getting a JS error:
> 
> image

a56bd978e should fix that.

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [inline assets/src/setup/pages/save/index.js:104]: Minor suggestion that I'm not totally sold on myself: since the admin bar gets hidden at `DOMContentLoaded` there is a flash of the admin bar. This could be avoided if we hide the `iframe` until the page is loaded.

```suggestion
					name="amp-wizard-completion-preview"
					style={ { opacity: 0 } }
					onLoad={ ( event ) => { event.target.style.opacity = '1'; } }
```

There's probably a more elegant way to do this however. (Including showing a spinner initially.) 

Not important.

**Comment by westonruter** [comment]
@johnwatkins0 When on the Done screen, the Close button and the Finish button do the same thing:

image

Should the close button be removed on the Done screen then? There's no sense to “Exit” the wizard when on the Done screen as the icon implies.

Also, I think the Close button should intentionally do `history.back()` to take the user to the screen they were just on (or else pass the return URL via a `return` query param similar to how the Customizer implements the Close button there).

**Update:** Filed as https://github.com/ampproject/amp-wp/issues/4974.

File: assets/src/setup/components/navigation-context-provider.js
Comment (westonruter) [inline assets/src/setup/components/navigation-context-provider.js:30]: Aside: I miss this query var 😄  It was useful to be on a screen and reload after changes to avoid having to tap Next Next Next.

**Comment by pierlon** [comment]
Should the text within the information panel on the template mode screen be selectable?

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js:104]: Done in 9b707af3265b3f67067f45029f5030ef8d2c8221

File: assets/src/setup/components/navigation-context-provider.js
Comment (johnwatkins0) [reply assets/src/setup/components/navigation-context-provider.js:30]: Should I put it back? I thought I was probably the only person using it.

**Comment by johnwatkins0** [comment]
> Should the text within the information panel on the template mode screen be selectable?
> 
> 

Updated in 8c320a6857170fe4a63d2105e428531fe91e4bd3. I think we can just let these be button-like because they will be genuine buttons in a few weeks.

File: assets/src/setup/pages/summary/test/redirect-toggle.js
Comment (westonruter) [inline assets/src/setup/pages/summary/test/redirect-toggle.js]:
```
/home/travis/build/ampproject/amp-wp/assets/src/setup/pages/summary/test/redirect-toggle.js
  22:10  error  'waitASecond' is defined but never used  no-unused-vars
```

Comment (johnwatkins0) [reply assets/src/setup/pages/summary/test/redirect-toggle.js]: Sorry, debug code. I've removed it.

**Review by westonruter — APPROVED** [review]
Functionality checks out. I'll merge once @pierlon gives his +1 on technical specifics.

**Review by pierlon — APPROVED** [review]
Looks good, ship it!

File: assets/src/setup/components/navigation-context-provider.js
Comment (westonruter) [reply assets/src/setup/components/navigation-context-provider.js:30]: Yes, please do. It's nice for development, no?

**Comment by schlessera** [comment]
>

**Comment by schlessera** [comment]
> This may be because you still have "classic" set as the reader theme, which won't be accepted by the validate_callback now that we've switched that slug to legacy.

Won't other folks have this as well? We'll probably need to ensure whatever value is found is handled gracefully (by silently setting to the correct default value, for example?).

**Comment by pierlon** [comment]
@schlessera I think he meant the [`reader_theme`](https://github.com/ampproject/amp-wp/blob/4139b2de3da1b036c5cc1c42bdebe2c4df07e2a1/includes/options/class-amp-options-manager.php#L34) option, which is not in production as yet.

**Pull request merged**
